keychain_apple.mm

```diff
@@ -100,31 +102,33 @@ static ErrorDescription fromStatus(OSStatus status)
             (__bridge id) kSecAttrAccount: (__bridge NSString *) key.toCFString(),
     };
 
-    OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef) query, nil);
+    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
+        OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef) query, nil);
 
-    if (status == errSecSuccess) {
-        NSDictionary *const update = @{
-                (__bridge id) kSecValueData: (__bridge NSData *) data.toCFData(),
-        };
+        if (status == errSecSuccess) {
+            NSDictionary *const update = @{
+                    (__bridge id) kSecValueData: (__bridge NSData *) data.toCFData(),
+            };
 
-        status = SecItemUpdate((__bridge CFDictionaryRef) query, (__bridge CFDictionaryRef) update);
-    } else {
-        NSDictionary *const insert = @{
-                (__bridge id) kSecClass: (__bridge id) kSecClassGenericPassword,
-                (__bridge id) kSecAttrService: (__bridge NSString *) service.toCFString(),
-                (__bridge id) kSecAttrAccount: (__bridge NSString *) key.toCFString(),
-                (__bridge id) kSecValueData: (__bridge NSData *) data.toCFData(),
-        };
+            status = SecItemUpdate((__bridge CFDictionaryRef) query, (__bridge CFDictionaryRef) update);
+        } else {
+            NSDictionary *const insert = @{
+                    (__bridge id) kSecClass: (__bridge id) kSecClassGenericPassword,
+                    (__bridge id) kSecAttrService: (__bridge NSString *) service.toCFString(),
+                    (__bridge id) kSecAttrAccount: (__bridge NSString *) key.toCFString(),
+                    (__bridge id) kSecValueData: (__bridge NSData *) data.toCFData(),
+            };
 
-        status = SecItemAdd((__bridge CFDictionaryRef) insert, nil);
-    }
+            status = SecItemAdd((__bridge CFDictionaryRef) insert, nil);
+        }
 
-    if (status == errSecSuccess) {
-        q->emitFinished();
-    } else {
-        const ErrorDescription error = ErrorDescription::fromStatus(status);
-        q->emitFinishedWithError(error.code,  tr("Could not store data in settings: %1").arg(error.message));
-    }
+        if (status == errSecSuccess) {
+            q->emitFinished();
```

Comment:
I don't think this is thread-safe, as this now seems to called in another thread than the main thread. The main thread would somehow need to be notified of the result and then send out the result signals.

Reply:
Hi @frankosterfeld you are right, so I have decided to rewrite the way the Apple keychain works internally. The summary is:

- There is a new Obj-C class called `AppleKeychainInterface` that responds to notifications emitted when the keychain operations are finished. This class takes pointers to relevant `Job` and `JobPrivate` objects that it is meant to operate (mainly by emitting the finished signals when it receives notifications that the keychain task is complete)
- The `JobPrivate` classes' `scheduledStart` methods now call global Obj-C functions that do the async keychain operations. These functions take in the required data for completing the keychain tasks as well as a `notificationId` (just an incrementing number) so that the `AppleKeychainInterface` can ignore notifications not relevant to it

This should make this thread-safe

Reply:
Now the testclient store/restore never terminates for me. I run e.g. `testclient store foo bar`, and I get asked for authorization the first time, but then nothing happens. (and nothing at all after the first time with the dialog)

Reply:
Fixed the `testclient`, turns out we need to do some init work to get the `NSNotificationCenter` to work. 

Thankfully, this is already done by `QGuiApplication`, hence why it worked in my test app and not the `testclient`. For the sake of simplicity have added `ifdef`s to use `QGuiApplication` rather than `QCoreApplication` for the `testclient` on Apple devices only, hope this is ok

Reply:
Ping? :)